invoice/src/main/java/org/killbill/billing/invoice/InvoiceDispatcher.java

```diff
@@ -572,90 +607,154 @@ private InvoiceWithFutureNotifications processAccountWithLockAndInputTargetDate(
-                }
+            final List<DefaultInvoice> updatedInvoices = invoicePluginDispatcher.updateOriginalInvoiceWithPluginInvoiceItems(originalInvoice, isDryRun, callContext, ImmutableList.<PluginProperty>of(), internalCallContext);
+            if (updatedInvoices == null) {
+                // Optimization
+                generatedInvoices.add(originalInvoice);
+            } else {
+                generatedInvoices.addAll(updatedInvoices);
+            }
 
-                // Use credit after we call the plugin (https://github.com/killbill/killbill/issues/637)
-                final InvoiceItem cbaItemPostInvoicePlugins = computeCBAOnExistingInvoice(invoice, internalCallContext);
-                if (cbaItemPostInvoicePlugins != null) {
-                    invoice.addInvoiceItem(cbaItemPostInvoicePlugins);
+            // Remove the temporary CBA item, unless it has been updated by the plugin
+            if (cbaItemPreInvoicePlugins != null) {
+                // We need to go through all invoices, in case the plugin decided to move that item around
+                invoiceDispatcher:
+                for (final Iterator<DefaultInvoice> invoiceIterator = generatedInvoices.iterator(); invoiceIterator.hasNext(); ) {
+                    final DefaultInvoice generatedInvoice = invoiceIterator.next();
+                    for (final Iterator<InvoiceItem> invoiceItemIterator = generatedInvoice.getInvoiceItems().iterator(); invoiceItemIterator.hasNext(); ) {
+                        final InvoiceItem generatedInvoiceItem = invoiceItemIterator.next();
+                        if (cbaItemPreInvoicePlugins.getId().equals(generatedInvoiceItem.getId()) &&
+                            // The plugin can override the description of the CBA item
+                            cbaItemPreInvoicePlugins.getDescription().equals(generatedInvoiceItem.getDescription())) {
+                            invoiceItemIterator.remove();
+                            if (generatedInvoice.getInvoiceItems().isEmpty()) {
+                                invoiceIterator.remove();
+                            }
+                            break invoiceDispatcher;
+                        }
+                    }
                 }
             }
 
-            if (!isDryRun) {
-                // Compute whether this is a new invoice object (or just some adjustments on an existing invoice), and extract invoiceIds for later use
-                final Set<UUID> uniqueInvoiceIds = getUniqueInvoiceIds(invoice);
-                final boolean isRealInvoiceWithItems = uniqueInvoiceIds.remove(invoice.getId());
-                final Set<UUID> adjustedUniqueOtherInvoiceId = uniqueInvoiceIds;
-
-                logInvoiceWithItems(account, invoice, actualTargetDate, adjustedUniqueOtherInvoiceId, isRealInvoiceWithItems);
+            // For non dry-run scenarii, the CBA will be recomputed in the DAO layer (https://github.com/killbill/killbill/issues/637)
```

Comment:
... unless description has changed and then DAO would recompute the missing part (e.g TAX) as needed.

Reply:
Happy to improve the comment, but I'm not sure I follow your suggestion.

At this point, the temporary CBA item has been removed if needed (i.e. if the description wasn't changed). Regardless, we need to recompute the CBA. This is either done manually (dry-run scenario) or automatically in the transaction (DAO layer): this is what this comment refers to (and why there is a `if` around the CBA call).